Update docs to use dbcp + webapp-runner + session service w/o docker

inodb opened the pull request:
Use dbcp + webapp-runner as default to deploy cBioPortal. Mention how to deploy session service w/o docker since it's a requirement for 3.0.0 and document the use of `-D` parameters.

Fix https://github.com/cBioPortal/cbioportal/issues/4832

I think using webapp-runner makes deploying the backend much easier compared to the Tomcat setup. No messing with Tomcat config files. I find `dbcp` a lot more clear than `jndi` as well. Only need to set database parameters in `portal.properties`. This is the exact same setup we are using on AWS (https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/cbioportal/cbioportal_spring_boot.yaml#L117). It makes sense for the default to be as similar to our own production environment as possible.

inodb commented on docs/Deploying.md:
````diff
@@ -21,108 +9,55 @@ Use it as a template to create your own:
+## Run the app
 
-``` xml
-<Context>
````
all these settings are already set when using dbcp

khzhu replied:
we do not need to configure maven to download webapp runner?

khzhu replied:
we do not need to configure mvn to download webapp runner?

inodb replied:
It is downloaded automatically (it's an existing dependency in the pom)

khzhu replied:
I see. great!

inodb commented on docs/Authenticating-Users-via-LDAP.md:
```diff
@@ -73,7 +69,14 @@ Rebuild the WAR file and re-deploy:
+    -jar \
+    portal/target/dependency/webapp-runner.jar \
+    --path /cbioportal \
+    portal/target/cbioportal.war
```
you can deploy on localhost:8080 instead of localhost:8080/cbioportal when not using `--path`. This might be better as default but I thought this would avoid breaking existing docs

fedde-s replied:
We've definitely been considering that as well, but not yet made a pull request for that same reason—nor documented it downstream in [thehyve/cbioportal-docker](https://github.com/thehyve/cbioportal-docker).

inodb replied:
@fedde-s maybe when i update the docker documentation i can do a find and replace of `localhost:8080/cbioportal` to `localhost:8080` (figured to do that next).

inodb replied:
i updated the path in this PR

fedde-s commented:
This is a major change in the deployment procedure. Shouldn't it go to `rc`?

fedde-s commented:
@inodb, you seem to have overwritten your commit while I was in the middle of a review. What did you change?

fedde-s commented:
I see—probably a rebase to `rc`.

khzhu commented on docs/Deploying.md:
````diff
@@ -21,108 +9,55 @@ Use it as a template to create your own:
+    portal/target/dependency/webapp-runner.jar \
+    --path /cbioportal \
+    portal/target/cbioportal.war
 ```
````
can we still configure 
1. the maximum number of active connections that can be allocated from this pool
2. the maximum number of connections that can remain idle in the pool, without extra ones being released,
since default values are set very small?

inodb replied:
It's all set here for the `dbcp` connection: https://github.com/cBioPortal/cbioportal/blob/fbd69216311e484947f55e30b180ba7244b6d824/core/src/main/java/org/mskcc/cbio/portal/dao/JdbcDataSource.java#L26-L36. We currently don't provide a way to change these, but we set them to what we were using in production

khzhu replied:
a good job, all configured already!!

khzhu replied:
if we can add some new properties in portal.properties to allow users to adjust those numbers that would be really helpful.

inodb replied:
I'm curious - did you play around with these values on your installation @khzhu or @fedde-s ? I imagine most people copy+pasted what we had in the docs. Feels a bit silly to introduce config that nobody will use

khzhu replied:
I do, when people complains the portal is too slow or take a long time to get response. As we move to clouds, with those advance performance monitor tools those numbers can be fine tuned or explored :) .

inodb replied:
@khzhu do you have any insights already to share on this? We haven't touched much of the mysql configuration on our end

khzhu replied:
it is a task we've given to Airgate, a professional cloud service firm Azure hired, to help come up with a configuration that best leverages the AKS we have. We will share our experience once it is done.

inodb replied:
i created a follow up issue for this https://github.com/cBioPortal/cbioportal/issues/6118

khzhu commented on docs/Authenticating-and-Authorizing-Users-via-keycloak.md:
```diff
@@ -245,7 +238,14 @@ You are now ready to go. Rebuild the WAR file and re-deploy:
+    -jar \
+    portal/target/dependency/webapp-runner.jar \
+    --path /cbioportal \
+    portal/target/cbioportal.war
```
agree. we should provide instructions on how to stop/restart the app.

inodb replied:
@fedde-s 

1. I can change the defaults to do a publicly accessible deploy with dbcp. That would remove the `-D` parameters from some of the commands. Also once I update the docker docs I can at the same time remove the `--path` parameter. That would remove another two lines. I'm not entirely sure what you mean by automate? Make a bash script?
2. If something's already running you get a "port in use" error. I can mention that. Do you think I should explain how to use ps, grep and kill to remove the process? Or provide some stop bash script. In addition, do you think there should be documentation on how to run this as a background process?

fedde-s replied:
1. Yes, making some of the ‘boilerplate’ arguments defaults may help, but I think the real solution would be highighting the settings to adjust _outside_ of the rest of the command. Pieter's request to list them as properties to add to (or change in) your portal.properties file would address this.
2. Maybe the real issue is that this is not a command to ‘re-deploy’; it is a command to start the server with some of the parameters for this use-case iff it isn't running yet. Thinking about it, do you think it makes sense to remove the various copies of the command from these pages entirely, and instead have them all link to the main guide on how to deploy? (`docs/Deploying.md`)

inodb replied:
@fedde-s Thanks for your reply!

1. Good point. I'll only leave the `-Dauthorization` then, b/c of https://github.com/cBioPortal/cbioportal/issues/6109
2. That's a good idea. Will try that

inodb replied:
This is done now

inodb commented on docs/Deploying.md:
this is not an issue with dbcp, because we set that here: https://github.com/cBioPortal/cbioportal/blob/fbd69216311e484947f55e30b180ba7244b6d824/core/src/main/java/org/mskcc/cbio/portal/dao/JdbcDataSource.java#L26-L36

inodb commented on docs/System-Requirements.md:
```diff
@@ -1,10 +1,18 @@
+instance. The public database consumes ~50 GB of disk space. If you do not plan
+to import public studies, depending on the size of your private data, 10GB of
+disk space may be sufficient.
 
```
@khzhu sorry, could you clarify? The system requirements should be the same on-premise? I can detail the spec of the AWS machine some more, but 32GB of RAM and 4 cores should suffice to run public portal. This is a better indication than we had with our old on-premise machine, cause it ran many instances of cBioPortal and therefore needed more resources (in the docs it mentions 192GB RAM)

khzhu replied:
@inodb , sorry for the confusion. It will be nice, if you can summarize in a table structure, something like:

> 
| Platform  | CPUs  | RAM |  Storage |
| ------------- | ------------- | ------------- | ------------- |
| on-premise  | 4  | 32  |  50 |
| aws  | 4 | 32  | 50  |

inodb replied:
put this in a new section `Hardware-Requirements`

inodb commented on docs/portal.properties-Reference.md:
````diff
@@ -180,12 +180,17 @@ googleplus.consumer.secret=2jCfg4SPWdGfXF44WC588dK
 ```sh
-CATALINA_OPTS='-Dauthenticate=googleplus'
+java \
+    -Dauthorization=false \
````
You are right. This documentation might need to be extended a bit, since you can also use google authentication with social_auth. That is, it is not required to have access, but it allows you to save user data such as sessions and virtual cohorts

inodb replied:
updated

pieterlukasse commented on docs/Authenticating-Users-via-SAML.md:
```diff
@@ -216,7 +211,14 @@ Rebuild the WAR file and re-deploy:
+java \
+    -Dauthorization=true \
+    -Dauthenticate=saml \
+    -Ddbconnector=dbcp \
```
@inodb what are the plans to get these properties from `portal.properties` instead? I feel we should actually try to get rid of all -D options in the example commands. IMO the documentation should tell users to adjust properties in the .properties file. We could then have only one section in documentation (for "advanced" users) telling how to overwrite portal.properties on runtime.

inodb replied:
@pieterlukasse that makes sense - I think all of these can go in `portal.properties` (those can btw also be overridden on runtime using `PORTAL_HOME` env variable). There was one issue with the `authorization` variable I think which required it to be set with `-D`. That's how it ended up in the documentation before (using `CATALINA_OPTS`). Maybe @fedde-s can elaborate what the issue was there

fedde-s replied:
True; for the `authenticate` variable, cBioPortal cannot currently use portal.properties at runtime. When implementing runtime configuration of auth in PR https://github.com/cBioPortal/cbioportal/pull/4917, none of us knew how to set up the Spring configuration (profile selection in XML files) to support that. That should be tracked as a separate issue, not introduced by this PR here.

For all other properties, I agree that we should document them in a consistent way, in portal.properties. Highlighting the specific settings to change for the use-case documented on this page from the ‘boilerplate code’ of the command would also address the issue I pointed out in my other review comment.

inodb replied:
@pieterlukasse @fedde-s agree with these comments. Let's only use `portal.properties`. With the exception for `authenticate`. Created a separate issue for that: https://github.com/cBioPortal/cbioportal/issues/6109.

As a side note, personally I prefer command line arguments over config files if there are proper `--help` / `man` docs, but I might be in the minority there. Regardless we are v far away from that, so using `properties` file def seems more feasible.

inodb commented:
@fedde-s @khzhu @pieterlukasse thanks a ton for the reviews! I think I've addressed most of the comments. I created a few separate issues to follow up on:

#6118
#6109

I'm planning to merge this soon - I can follow up with new PRs

inodb commented:
Also this one: https://github.com/cBioPortal/cbioportal/issues/6120 (no longer necessary to compile the source code, we could download jar/war directly)

fedde-s commented:
You seem to have accidentally hit the merge button halfway through the review process; you should probably revert the PR.

inodb commented:
@fedde-s sorry we figured it to be better to merge some of this documentation and create and address any remaining things separately. But I can revert it if preferred. Lmk if there are any specific concerns

fedde-s commented on src/main/resources/portal.properties.EXAMPLE:
```diff
@@ -8,7 +8,10 @@ db.host=localhost
 db.connection_string=jdbc:mysql://localhost/
-db.tomcat_resource_name=jdbc/cbioportal
+# set tomcat_resource_name when using dbconnector=jndi instead of the defaault
+# dbconnector=dbcp
```
@inodb, did you test this? I was under the impression that the `dbconnector` setting only worked if it was supplied as a `-D` property to the JVM, since it's being used as a Spring applicationContext.xml profile in the same way as the other `authenticate` setting listed in  https://github.com/cBioPortal/cbioportal/issues/6109

fedde-s replied:
Never mind, I see you're already tracking this as a followup task in https://github.com/cBioPortal/cbioportal/issues/6148
